File: Thanh/ServerEmbedding/src/util/connectmysql.py
import mysql.connector
from mysql.connector import Error

class ConnectDB:
    def __init__(self):
        self.connection = self.get_connection()
        self.cursor = self.connection.cursor()

    def __del__(self):
        self.connection.close()
        print("MySQL connection is closed")

    def get_connection(self):
        try:
            connection = mysql.connector.connect(host='localhost',
                                                 database='EmbServer',
                                                 user='root',
                                                 password='31072001')
            if connection.is_connected():
                db_Info = connection.get_server_info()
                # print("Connected to MySQL Server version ", db_Info)
                cursor = connection.cursor()
                cursor.execute("select database();")
                record = cursor.fetchone()
                print("Your connected to database: ", record)
            return connection
        except Error as e:
            print("Error while connecting to MySQL", e)

    def create_table_controll(self):

        try:
            self.__init__()
            sql = """CREATE TABLE IF NOT EXISTS Controll (
                    id INT(11) NOT NULL AUTO_INCREMENT,
                    theta_1 FLOAT(11) NOT NULL,
                    theta_2 FLOAT(11) NOT NULL,
                    theta_3 FLOAT(11) NOT NULL,
                    theta_4 FLOAT(11) NOT NULL,
                    theta_5 FLOAT(11) NOT NULL,
                    w_1 FLOAT(11) NOT NULL,
                    w_2 FLOAT(11) NOT NULL,
                    w_3 FLOAT(11) NOT NULL,
                    w_4 FLOAT(11) NOT NULL,
                    w_5 FLOAT(11) NOT NULL,
                    vitri_x FLOAT(11) NOT NULL,
                    vitri_y FLOAT(11) NOT NULL,
                    vitri_z FLOAT(11) NOT NULL,
                    phi FLOAT(11) NOT NULL,
                    gramma FLOAT(11) NOT NULL,
                    v FLOAT(11) NOT NULL,
                    PRIMARY KEY (id)
                    )
                    ENGINE=InnoDB;"""
            self.cursor.execute(sql)
            self.connection.commit()
            print("Table created successfully")
            
        except Error as e:
            print("Error while creating table", e)
        finally:
            self.cursor.close()

    def create_table_motor_default(self):
        try:
            self.__init__()
            sql = """CREATE TABLE IF NOT EXISTS MotorDefault (
                    id INT(11) NOT NULL AUTO_INCREMENT,
                    theta_1 FLOAT(11) NOT NULL,
                    theta_2 FLOAT(11) NOT NULL,
                    theta_3 FLOAT(11) NOT NULL,
                    theta_4 FLOAT(11) NOT NULL,
                    theta_5 FLOAT(11) NOT NULL,
                    w_1 FLOAT(11) NOT NULL,
                    w_2 FLOAT(11) NOT NULL,
                    w_3 FLOAT(11) NOT NULL,
                    w_4 FLOAT(11) NOT NULL,
                    w_5 FLOAT(11) NOT NULL,
                    vitri_x FLOAT(11) NOT NULL,
                    vitri_y FLOAT(11) NOT NULL,
                    vitri_z FLOAT(11) NOT NULL,
                    phi FLOAT(11) NOT NULL,
                    gramma FLOAT(11) NOT NULL,
                    v FLOAT(11) NOT NULL,
                    position FLOAT(11) NOT NULL,
                    velocity FLOAT(11) NOT NULL,
                    description VARCHAR(255) NOT NULL,
                    PRIMARY KEY (id)
                    )
                    ENGINE=InnoDB;"""
                    
            self.cursor.execute(sql)
            self.connection.commit()
            print("Table created successfully")
        except Error as e:
            print("Error while creating table", e)
        finally:
            self.cursor.close()

    def insert_data(self, data, table_name: str ="Controll", sql : str = None):
        try:
            self.__init__()
            if data:
                if type(data) != type(list):
                    data = list(data)
                if table_name == "Controll":
                    sql = """INSERT INTO Controll (theta_1, theta_2, theta_3, theta_4, theta_5, w_1, w_2, w_3, w_4, w_5, vitri_x, vitri_y, vitri_z, phi, gramma, v) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)"""
                elif table_name == "MotorDefault":
                    sql = """INSERT INTO MotorDefault (theta_1, theta_2, theta_3, theta_4, theta_5, w_1, w_2, w_3, w_4, w_5, vitri_x, vitri_y, vitri_z, phi, gramma, v, position, velocity, description) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)"""
                self.cursor.execute(sql, data)
            elif sql:
                self.cursor.execute(sql)
            
            self.connection.commit()
            print("Data inserted successfully")
        except Error as e:
            print("Error while inserting data", e)
        finally:
            self.cursor.close()

    def update_data(self, id, data, name_database='Controll'):
        if type(data) != type(list):
            data = list(data)
        try:
            self.__init__()
            if name_database == "Controll":
                sql = """UPDATE Controll SET theta_1 = %s, theta_2 = %s, theta_3 = %s, theta_4 = %s, theta_5 = %s, w_1 = %s, w_2 = %s, w_3 = %s, w_4 = %s, w_5 = %s, vitri_x = %s, vitri_y = %s, vitri_z = %s, phi = %s, gramma = %s, v = %s WHERE id = """ + str(id)
            elif name_database == "MotorDefault":
                sql = """UPDATE MotorDefault SET theta_1 = %s, theta_2 = %s, theta_3 = %s, theta_4 = %s, theta_5 = %s, w_1 = %s, w_2 = %s, w_3 = %s, w_4 = %s, w_5 = %s, vitri_x = %s, vitri_y = %s, vitri_z = %s, phi = %s, gramma = %s, v = %s, position = %s, velocity = %s, description = %s WHERE id = """ + str(id)
            self.cursor.execute(sql, data)
            self.connection.commit()
            print("Data updated successfully")
        except Error as e:
            print("Error while updating data", e)
        finally:
            self.cursor.close()

    def delete_data_by_id(self, id, name_database='Controll'):
        #  convert id to list
        id = str(id)
        if type(id) != type(list):
            id = list(id)
        try:
            self.__init__()
            if name_database == 'Controll':
                sql = """DELETE FROM Controll WHERE id = %s"""
            elif name_database == 'MotorDefault':
                sql = """DELETE FROM MotorDefault WHERE id = %s"""
            self.cursor.execute(sql, id)
            self.connection.commit()
            print("Data deleted successfully")
        except Error as e:
            print("Error while deleting data", e)
        finally:
            self.cursor.close()
    
    def delete_data_by_feature(self, feature, value):
        data = [feature, value]
        try:
            self.__init__()
            sql = """DELETE FROM Controll WHERE %s = %s"""
            self.cursor.execute(sql, data)
            self.connection.commit()
            print("Data deleted successfully")
        except Error as e:
            print("Error while deleting data", e)
        finally:
            self.cursor.close()

    def select_data(self):
        try:
            self.__init__()
            sql = """SELECT * FROM Controll"""
            self.cursor.execute(sql)
            result = self.cursor.fetchall()
            return result
        except Error as e:
            print("Error while selecting data", e)
        finally:
            self.cursor.close()

    def select_data_by_id(self, id):
        try:
            self.__init__()
            sql = """SELECT * FROM Controll WHERE id = %s"""
            self.cursor.execute(sql, id)
            result = self.cursor.fetchall()
            return result
        except Error as e:
            print("Error while selecting data", e)
        finally:
            self.cursor.close()

    def select_data_by_mode(self, mode):
        try:
            self.__init__()
            sql = """SELECT * FROM Controll WHERE mode = %s"""
            self.cursor.execute(sql, mode)
            result = self.cursor.fetchall()
            return result
        except Error as e:
            print("Error while selecting data", e)
        finally:
            self.cursor.close()

    def show_data(self, table_name="Controll"):
        try:
            # check if table exists
            self.__init__()
            sql = """SELECT * FROM %s""" % table_name
            self.cursor.execute(sql)
            result = self.cursor.fetchall()
            return result
        except Error as e:
            print("Error while showing data", e)
        finally:
            self.cursor.close()

# ConnectDB = ConnectDB()
# ConnectDB.create_table_controll()
# ConnectDB.create_table_motor_default()
# result = ConnectDB.show_data()
# print(result)
# # insert data into table Controll
# data = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, "mode2", 1, 2, 3, "position2", 'velocity2']
# data = [10, 20, 30, 40, 50, 60, 70, 80, 90, 100, "mode1", 1, 2, 3, "position1", 'velocity1']
# ConnectDB.insert_data(data, table_name="MotorDefault")
# data = [100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, "mode3", 1, 2, 3, "position3", 'velocity3']
# ConnectDB.insert_data(data, table_name="MotorDefault")

File: Thanh/ServerEmbedding/src/util/connectmysql.c
#include <stdio.h>
#include <mysql/mysql.h>

// function connect mysql return connection
MYSQL *connect_mysql() {
  MYSQL *conn = NULL;
  conn = mysql_init(NULL);
  if (conn == NULL) {
    fprintf(stderr, "mysql_init() failed\n");
    return NULL;
  }
  if (mysql_real_connect(conn, "localhost", "root", "31072001", "mydatabase", 0, NULL, 0) == NULL) {
    fprintf(stderr, "mysql_real_connect() failed\n");
    mysql_close(conn);
    return NULL;
  }
  return conn;
}

// create table if not exists motor default
int create_table_motor_default(MYSQL *conn) {
  if (mysql_query(conn, "create table IF NOT EXISTS MotorDefault (\
                    id INT(11) NOT NULL AUTO_INCREMENT,\
                    theta_1 FLOAT(11) NOT NULL,\
                    theta_2 FLOAT(11) NOT NULL,\
                    theta_3 FLOAT(11) NOT NULL,\
                    theta_4 FLOAT(11) NOT NULL,\
                    theta_5 FLOAT(11) NOT NULL,\
                    w_1 FLOAT(11) NOT NULL,\
                    w_2 FLOAT(11) NOT NULL,\
                    w_3 FLOAT(11) NOT NULL,\
                    w_4 FLOAT(11) NOT NULL,\
                    w_5 FLOAT(11) NOT NULL,\
                    vitri_x FLOAT(11) NOT NULL,\
                    vitri_y FLOAT(11) NOT NULL,\
                    vitri_z FLOAT(11) NOT NULL,\
                    phi FLOAT(11) NOT NULL,\
                    gramma FLOAT(11) NOT NULL,\
                    v FLOAT(11) NOT NULL,\
                    position FLOAT(11) NOT NULL,\
                    velocity FLOAT(11) NOT NULL,\
                    description VARCHAR(255) NOT NULL,\
                    PRIMARY KEY (id)\
                    )\
                    ENGINE=InnoDB;")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  printf("Connected to database\n");
  return 0;
}

// create table if not exists motor
int create_table_controll(MYSQL *conn) {
  if (mysql_query(conn, "create table IF NOT EXISTS Controll (\
                    id INT(11) NOT NULL AUTO_INCREMENT,\
                    theta_1 FLOAT(11) NOT NULL,\
                    theta_2 FLOAT(11) NOT NULL,\
                    theta_3 FLOAT(11) NOT NULL,\
                    theta_4 FLOAT(11) NOT NULL,\
                    theta_5 FLOAT(11) NOT NULL,\
                    w_1 FLOAT(11) NOT NULL,\
                    w_2 FLOAT(11) NOT NULL,\
                    w_3 FLOAT(11) NOT NULL,\
                    w_4 FLOAT(11) NOT NULL,\
                    w_5 FLOAT(11) NOT NULL,\
                    vitri_x FLOAT(11) NOT NULL,\
                    vitri_y FLOAT(11) NOT NULL,\
                    vitri_z FLOAT(11) NOT NULL,\
                    phi FLOAT(11) NOT NULL,\
                    gramma FLOAT(11) NOT NULL,\
                    v FLOAT(11) NOT NULL,\
                    PRIMARY KEY (id)\
                    )\
                    ENGINE=InnoDB;")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  printf("Connected to database\n");
  return 0;
}

// insert data to table motor
int insert_data_motor(MYSQL *conn, float theta_1, float theta_2, float theta_3, float theta_4, float theta_5, float w_1, float w_2, float w_3, float w_4, float w_5, float vitri_x, float vitri_y, float vitri_z, float phi, float gramma, float v, float position, float velocity, char *description) {
  if (mysql_query(conn, "insert into MotorDefault (theta_1, theta_2, theta_3, theta_4, theta_5, w_1, w_2, w_3, w_4, w_5, vitri_x, vitri_y, vitri_z, phi, gramma, v, position, velocity, description) values ('" + theta_1 + "', '" + theta_2 + "', '" + theta_3 + "', '" + theta_4 + "', '" + theta_5 + "', '" + w_1 + "', '" + w_2 + "', '" + w_3 + "', '" + w_4 + "', '" + w_5 + "', '" + vitri_x + "', '" + vitri_y + "', '" + vitri_z + "', '" + phi + "', '" + gramma + "', '" + v + "', '" + position + "', '" + velocity + "', '" + description + "')")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  printf("Insert data to table motor\n");
  return 0;
}

// insert data to table controll
int insert_data_controll(MYSQL *conn, float theta_1, float theta_2, float theta_3, float theta_4, float theta_5, float w_1, float w_2, float w_3, float w_4, float w_5, float vitri_x, float vitri_y, float vitri_z, float phi, float gramma, float v) {
  if (mysql_query(conn, "insert into Controll (theta_1, theta_2, theta_3, theta_4, theta_5, w_1, w_2, w_3, w_4, w_5, vitri_x, vitri_y, vitri_z, phi, gramma, v) values ('" + theta_1 + "', '" + theta_2 + "', '" + theta_3 + "', '" + theta_4 + "', '" + theta_5 + "', '" + w_1 + "', '" + w_2 + "', '" + w_3 + "', '" + w_4 + "', '" + w_5 + "', '" + vitri_x + "', '" + vitri_y + "', '" + vitri_z + "', '" + phi + "', '" + gramma + "', '" + v + "')")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  printf("Insert data to table controll\n");
  return 0;
}

// select data from table motor and return data
int select_data_motor(MYSQL *conn, float *theta_1, float *theta_2, float *theta_3, float *theta_4, float *theta_5, float *w_1, float *w_2, float *w_3, float *w_4, float *w_5, float *vitri_x, float *vitri_y, float *vitri_z, float *phi, float *gramma, float *v, float *position, float *velocity, char *description) {
  MYSQL_RES *res;
  MYSQL_ROW row;
  if (mysql_query(conn, "select * from MotorDefault")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  res = mysql_store_result(conn);
  if (res == NULL) {
    fprintf(stderr, "mysql_store_result() failed\n");
    return -1;
  }
  row = mysql_fetch_row(res);
  if (row == NULL) {
    fprintf(stderr, "mysql_fetch_row() failed\n");
    return -1;
  }
  *theta_1 = atof(row[1]);
    *theta_2 = atof(row[2]);
    *theta_3 = atof(row[3]);
    *theta_4 = atof(row[4]);
    *theta_5 = atof(row[5]);
    *w_1 = atof(row[6]);
    *w_2 = atof(row[7]);
    *w_3 = atof(row[8]);
    *w_4 = atof(row[9]);
    *w_5 = atof(row[10]);
    *vitri_x = atof(row[11]);
    *vitri_y = atof(row[12]);
    *vitri_z = atof(row[13]);
    *phi = atof(row[14]);
    *gramma = atof(row[15]);
    *v = atof(row[16]);
    *position = atof(row[17]);
    *velocity = atof(row[18]);
    strcpy(description, row[19]);
    return 0;
}

// select data from table controll and return data to controll
int select_data_controll(MYSQL *conn, float *theta_1, float *theta_2, float *theta_3, float *theta_4, float *theta_5, float *w_1, float *w_2, float *w_3, float *w_4, float *w_5, float *vitri_x, float *vitri_y, float *vitri_z, float *phi, float *gramma, float *v) {
  MYSQL_RES *res;
  MYSQL_ROW row;
  if (mysql_query(conn, "select * from Controll")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  res = mysql_store_result(conn);
  if (res == NULL) {
    fprintf(stderr, "mysql_store_result() failed\n");
    return -1;
  }
  row = mysql_fetch_row(res);
  if (row == NULL) {
    fprintf(stderr, "mysql_fetch_row() failed\n");
    return -1;
  }
  *theta_1 = atof(row[1]);
    *theta_2 = atof(row[2]);
    *theta_3 = atof(row[3]);
    *theta_4 = atof(row[4]);
    *theta_5 = atof(row[5]);

    *w_1 = atof(row[6]);
    *w_2 = atof(row[7]);
    *w_3 = atof(row[8]);
    *w_4 = atof(row[9]);
    *w_5 = atof(row[10]);
    *vitri_x = atof(row[11]);
    *vitri_y = atof(row[12]);
    *vitri_z = atof(row[13]);
    *phi = atof(row[14]);
    *gramma = atof(row[15]);
    *v = atof(row[16]);
    return 0;
}

// update data to table motor
int update_data_motor(MYSQL *conn, float theta_1, float theta_2, float theta_3, float theta_4, float theta_5, float w_1, float w_2, float w_3, float w_4, float w_5, float vitri_x, float vitri_y, float vitri_z, float phi, float gramma, float v, float position, float velocity, char *description) {
  if (mysql_query(conn, "update MotorDefault set theta_1 = '" + theta_1 + "', theta_2 = '" + theta_2 + "', theta_3 = '" + theta_3 + "', theta_4 = '" + theta_4 + "', theta_5 = '" + theta_5 + "', w_1 = '" + w_1 + "', w_2 = '" + w_2 + "', w_3 = '" + w_3 + "', w_4 = '" + w_4 + "', w_5 = '" + w_5 + "', vitri_x = '" + vitri_x + "', vitri_y = '" + vitri_y + "', vitri_z = '" + vitri_z + "', phi = '" + phi + "', gramma = '" + gramma + "', v = '" + v + "', position = '" + position + "', velocity = '" + velocity + "', description = '" + description + "'")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  printf("Update data to table motor\n");
  return 0;
}

// update data to table controll
int update_data_controll(MYSQL *conn, float theta_1, float theta_2, float theta_3, float theta_4, float theta_5, float w_1, float w_2, float w_3, float w_4, float w_5, float vitri_x, float vitri_y, float vitri_z, float phi, float gramma, float v) {
  if (mysql_query(conn, "update Controll set theta_1 = '" + theta_1 + "', theta_2 = '" + theta_2 + "', theta_3 = '" + theta_3 + "', theta_4 = '" + theta_4 + "', theta_5 = '" + theta_5 + "', w_1 = '" + w_1 + "', w_2 = '" + w_2 + "', w_3 = '" + w_3 + "', w_4 = '" + w_4 + "', w_5 = '" + w_5 + "', vitri_x = '" + vitri_x + "', vitri_y = '" + vitri_y + "', vitri_z = '" + vitri_z + "', phi = '" + phi + "', gramma = '" + gramma + "', v = '" + v + "'")) {
    fprintf(stderr, "mysql_query() failed\n");
    return -1;
  }
  printf("Update data to table controll\n");
  return 0;
}

// test main
int main() {
  connect_mysql();
  
  mysql_close(conn);

    return 0;
}
